Networking/Networking/ApplicationPassword/RequestAuthenticator.swift

```diff
@@ -0,0 +1,91 @@
+enum RequestAuthenticatorError: Error {
+    case applicationPasswordUseCaseNotAvailable
+    case applicationPasswordNotAvailable
+}
+
+/// Authenticates request
+///
+public struct RequestAuthenticator {
+    /// Credentials.
+    ///
+    let credentials: Credentials?
+
+    /// The use case to handle authentication with application passwords.
+    ///
+    private let applicationPasswordUseCase: ApplicationPasswordUseCase?
+
+    /// Sets up the authenticator with optional credentials and application password use case.
+    /// `applicationPasswordUseCase` can be injected for unit tests.
+    ///
+    init(credentials: Credentials?, applicationPasswordUseCase: ApplicationPasswordUseCase? = nil) {
+        self.credentials = credentials
+        let useCase: ApplicationPasswordUseCase? = {
+            if let applicationPasswordUseCase {
+                return applicationPasswordUseCase
+            } else if case let .wporg(username, password, siteAddress) = credentials {
+                return try? DefaultApplicationPasswordUseCase(username: username,
+                                                              password: password,
+                                                              siteAddress: siteAddress)
```

Comment:
Not directly related to this PR: I took a look at what might cause the use case to throw an error, and it looks like it could be from an invalid site address `ApplicationPasswordUseCaseError.invalidSiteAddress`. Can we validate the site address when the user signs in with site credentials so that the use case can be non-nil here?

Reply:
The `DefaultApplicationPasswordUseCase` initialization can throw an error when we are not able to construct login/admin URL using the site address, [here](https://github.com/woocommerce/woocommerce-ios/pull/8483/files#diff-6f412183a8c153584c372b45458f65ddc92da5bdd69e8d90d63463af4298d5c1R77-R78)

I have renamed the error in this commit to explain the scenario better. ae5b15c9df4d4810bcf37786d7efc6c82b80c444

> Can we validate the site address when the user signs in with site credentials so that the use case can be non-nil here?

Sorry I don't understand this. Could you kindly explain how we can improve it? Thank you!

Reply:
During the login flow where the user signs in with site credentials, do we check if the login and admin URLs are valid? If so, maybe we don't need to check again so that the initialization always succeeds. I assume the API requests here are after the login flow, please correct me if this isn't the case.

Reply:
We validate `loginURL` and `adminURL` only while sending `WordPressOrgNetwork`, which currently happens during Jetpack setup [here](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Authentication/Jetpack%20Setup/Site%20Credential%20Login/SiteCredentialLoginViewModel.swift#L81-L86). So we need to validate the  `loginURL` and `adminURL` here again.

Reply:
> does it make sense to validate these two URLs during the login step so that we don't have to handle this error scenario here?

I agree. Created an issue https://github.com/woocommerce/woocommerce-ios/issues/8506